### DISCUSSION TITLE
2021/12/santa-sleigh-xunit/

### DISCUSSION THREAD

**Discussion by giscus** [SeanKilleen/seankilleen.github.io#730] [Specific Post Comments]
# 2021/12/santa-sleigh-xunit/

Solving our C# Advent 2021 challenge using the xUnit framework.

https://seankilleen.com/2021/12/santa-sleigh-xunit/

**Comment by isaacabraham** [comment]
Sweet!

As a quick exercise (and without realising you'd used FSCheck already ;-)) I had a quick bash to identify some properties which would reduce the number of tests you had for the basic elements. I came up with a few properties here:

https://gist.github.com/isaacabraham/e74ccb69cee86112eee7148ec59d27e4#file-santa-fs-L44

Basically things like:

* Moving *always* gives a different position to where you started
* If you move back and forward you're back where you started
* Moving in two different directions always gives a different result

etc.

Might give you some ideas for some opportunities to trim some of those tests you have even further :-)

**Reply by SeanKilleen** [reply]
Nice! Love the F# translation as well :) 

My particular experience has been that it's helpful for folks new to automated testing to be explicit about what they're looking for in scenarios, and then refactor to property-based tests later, especially when they're newer to testing. In the classes I've taught, I found that a lot of people tend to get in over their head with property-based tests and confuse themselves, or might write tests that don't do exactly what they think they do. 

Still, this is a very valid comment a great additional resource for readers. Thanks for taking the time!

**Reply by isaacabraham** [reply]
Agreed with all that :-) I think I often tend to look for PBTs when I end up with lots of hand-written unit tests that have arbitrary inputs. They're all good options to have - and PBTs are definitely tricky to get started with - there's much less literature on how to go about identifying possible tests than conventional unit tests.
